metadata/src/test/java/org/apache/kafka/image/publisher/BrokerRegistrationTrackerTest.java

```diff
@@ -0,0 +1,125 @@
+
+    static class BrokerRegistrationTrackerTestContext {
+        AtomicInteger numCalls = new AtomicInteger(0);
+        BrokerRegistrationTracker tracker = new BrokerRegistrationTracker(1,
+                Arrays.asList(B, A), () -> numCalls.incrementAndGet());
+
+        MetadataImage image = MetadataImage.EMPTY;
+
+        void onMetadataUpdate(MetadataDelta delta) {
+            MetadataProvenance provenance = new MetadataProvenance(0, 0, 0);
+            image = delta.apply(provenance);
+            LogDeltaManifest manifest = new LogDeltaManifest.Builder().
+                provenance(provenance).
+                leaderAndEpoch(LeaderAndEpoch.UNKNOWN).
+                numBatches(1).
+                elapsedNs(1).
+                numBytes(1).
+                build();
+            tracker.onMetadataUpdate(delta, image, manifest);
+        }
+
+        MetadataDelta newDelta() {
+            return new MetadataDelta.Builder().
+                setImage(image).
+                build();
+        }
+    }
+
+    @Test
+    public void testTrackerName() {
+        BrokerRegistrationTrackerTestContext ctx  = new BrokerRegistrationTrackerTestContext();
+        assertEquals("BrokerRegistrationTracker(id=1)", ctx.tracker.name());
+    }
+
+    @Test
+    public void testMetadataVersionUpdateWithoutRegistrationDoesNothing() {
+        BrokerRegistrationTrackerTestContext ctx  = new BrokerRegistrationTrackerTestContext();
+        MetadataDelta delta = ctx.newDelta();
+        delta.replay(new FeatureLevelRecord().
+            setName(MetadataVersion.FEATURE_NAME).
+            setFeatureLevel(MetadataVersion.IBP_3_7_IV2.featureLevel()));
+        ctx.onMetadataUpdate(delta);
+        assertEquals(0, ctx.numCalls.get());
+    }
+
+    @Test
+    public void testBrokerUpdateWithoutNewMvDoesNothing() {
+        BrokerRegistrationTrackerTestContext ctx  = new BrokerRegistrationTrackerTestContext();
+        MetadataDelta delta = ctx.newDelta();
+        delta.replay(new RegisterBrokerRecord().
+            setBrokerId(1).
+            setIncarnationId(INCARNATION_ID).
+            setLogDirs(Arrays.asList(A, B, C)));
+        ctx.onMetadataUpdate(delta);
+        assertEquals(0, ctx.numCalls.get());
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    public void testBrokerUpdateWithNewMv(boolean jbodMv) {
+        BrokerRegistrationTrackerTestContext ctx  = new BrokerRegistrationTrackerTestContext();
+        MetadataDelta delta = ctx.newDelta();
+        delta.replay(new RegisterBrokerRecord().
+            setBrokerId(1).
+            setIncarnationId(INCARNATION_ID).
+            setLogDirs(Arrays.asList()));
+        delta.replay(new FeatureLevelRecord().
+            setName(MetadataVersion.FEATURE_NAME).
+            setFeatureLevel(jbodMv ? MetadataVersion.IBP_3_7_IV2.featureLevel() :
+                MetadataVersion.IBP_3_7_IV1.featureLevel()));
+        ctx.onMetadataUpdate(delta);
```

Comment:
On a second read, I think my confusion stems from `onMetadataUpdate` being invoked with 2 deltas together in this test.

Perhaps the following conveys the motivation behind the if condition in `BrokerRegistrationTracker:78-86`?:

```java
        BrokerRegistrationTrackerTestContext ctx  = new BrokerRegistrationTrackerTestContext();
        MetadataDelta delta = ctx.newDelta();
        delta.replay(new RegisterBrokerRecord().
            setBrokerId(1).
            setIncarnationId(INCARNATION_ID).
            setLogDirs(Arrays.asList()));
        ctx.onMetadataUpdate(delta);
        // No calls are made because MetadataVersion is 3.0-IV1 initially
        assertEquals(0, ctx.numCalls.get());

        delta = ctx.newDelta();
        delta.replay(new FeatureLevelRecord().
            setName(MetadataVersion.FEATURE_NAME).
            setFeatureLevel(jbodMv ? MetadataVersion.IBP_3_7_IV2.featureLevel() :
                MetadataVersion.IBP_3_7_IV1.featureLevel()));
        ctx.onMetadataUpdate(delta);
        if (jbodMv) {
            assertEquals(1, ctx.numCalls.get());
        } else {
            assertEquals(0, ctx.numCalls.get());
        }
```

Reply:
Yes, this is a good test. I have added it in addition to the other ones.